pkg/manager/tidbngmonitoring/ng_monitoring_manager.go

```diff
@@ -0,0 +1,474 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tidbngmonitoring
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/label"
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/manager"
+	mngerutils "github.com/pingcap/tidb-operator/pkg/manager/utils"
+	"github.com/pingcap/tidb-operator/pkg/util"
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog"
+)
+
+const (
+	ngmPodDataVolumeMountDir   = "/var/lib/ng-monitoring" // the mount path for ng monitoring data volume
+	ngmPodConfigVolumeMountDir = "/etc/ng-monitoring"     // the dir for ng monitoring config
+	ngmPodConfigFilename       = "ng-monitoring.toml"     // the filename of config file
+
+	ngmConfigMapConfigKey = "ng-monitoring-config" // the key for config data in config map
+
+	ngmServicePort = 12020
+)
+
+type ngMonitoringManager struct {
+	deps *controller.Dependencies
+}
+
+func NewNGMonitorManager(deps *controller.Dependencies) manager.TiDBNGMonitoringManager {
+	return &ngMonitoringManager{
+		deps: deps,
+	}
+}
+
+func (m *ngMonitoringManager) Sync(tngm *v1alpha1.TiDBNGMonitoring) error {
+	var err error
+
+	err = m.syncService(tngm)
+	if err != nil {
+		return err
+	}
+
+	err = m.syncCore(tngm)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *ngMonitoringManager) syncService(tngm *v1alpha1.TiDBNGMonitoring) error {
+	ns := tngm.GetNamespace()
+	name := tngm.GetName()
+
+	if tngm.Spec.Paused {
+		klog.V(4).Infof("tikv cluster %s/%s is paused, skip syncing for ng monitoring headless service", ns, name)
+		return nil
+	}
+
+	newSvc := GenerateNGMonitoringHeadlessService(tngm)
+	oldSvc, err := m.deps.ServiceLister.Services(newSvc.Namespace).Get(newSvc.Name)
+	svcNotFound := errors.IsNotFound(err)
+
+	if err != nil && !svcNotFound {
+		return fmt.Errorf("syncService: failed to get headless svc %s/%s for ng monitoring %s/%s, error %s", newSvc.Namespace, newSvc.Name, ns, name, err)
+	}
+
+	// first creation
+	if svcNotFound {
+		err := controller.SetServiceLastAppliedConfigAnnotation(newSvc)
+		if err != nil {
+			return err
+		}
+		return m.deps.ServiceControl.CreateService(tngm, newSvc)
+	}
+
+	// update existing service if needed
+	equal, err := controller.ServiceEqual(newSvc, oldSvc)
+	if err != nil {
+		return err
+	}
+	isOrphan := metav1.GetControllerOf(oldSvc) == nil
+	if !equal || isOrphan {
+		svc := *oldSvc
+		svc.Spec = newSvc.Spec
+		err = controller.SetServiceLastAppliedConfigAnnotation(&svc)
+		if err != nil {
+			return err
+		}
+		// Adopt headless-service created by helm
```

Comment:
? can we remove the handler of `isOrphan` for ngm?

Reply:
removed